### PR TITLE
LOG-1283: Add linting and Dockerfile generation from midstream Docker.in

### PR DESCRIPTION
**Pull request by jcantrill**
### Description
This PR introduces a way to generate CI dockerfile from midstream dockerfile

/hold

cc @vimalk78 

https://issues.redhat.com/browse/LOG-1283

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,19 +1,25 @@
...
-COPY apis apis
-COPY controllers controllers
-COPY files files
-COPY internal internal
-COPY manifests manifests
-COPY vendor vendor
-COPY version version
-ADD Makefile main.go ./
-RUN go build -o bin/elasticsearch-operator main.go
+COPY ${REMOTE_SOURCE} .
```
Comment (ewolinetz) [inline Dockerfile:14]: we were intentionally not copying everything before... can you please fix?

File: Dockerfile.in [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ENV OS_GIT_PATCH=${CI_Z_VERSION}
+ENV SOURCE_GIT_COMMIT=${CI_ELASTICSEARCH_OPERATOR_UPSTREAM_COMMIT}
+ENV SOURCE_GIT_URL=${CI_ELASTICSEARCH_OPERATOR_UPSTREAM_URL}
+
+## EXCLUDE BEGIN ##
+ENV REMOTE_SOURCE=${REMOTE_SOURCE}/app
+## EXCLUDE END ##
+
+WORKDIR /go/src/github.com/openshift/elasticsearch-operator
+COPY ${REMOTE_SOURCE} .
```
Comment (ewolinetz) [inline Dockerfile.in:16]: same ask

**Review by ewolinetz — CHANGES_REQUESTED** [review]
request to preserve pattern we were using before for copying files before build

File: Dockerfile (hunk 1 above)
Comment (alanconway) [reply Dockerfile:14]: Curious - which parts are we not copying and why do they need to be in the src/ directory if they are not needed to build?

Comment (jcantrill) [reply Dockerfile]: modified

File: Makefile [hunk 3]
```diff
@@ -85,13 +85,9 @@ lint-dockerfile:
 	@hack/lint-dockerfile
 .PHONY: lint-dockerfile
 
-.INTERMEDIATE: Dockerfile.dev
-Dockerfile.dev: Dockerfile Dockerfile.centos.patch
```
Comment (ewolinetz) [inline Makefile:89]:
```
i will remove that task.  I don't see the point at the moment
```

can you revert this? we use it so we don't need to build using any of the registry images locally

File: Makefile [hunk 4]
```diff
@@ -236,4 +232,7 @@ elasticsearch-operator-install:
 # uninstalls the elasticsearch operator
 elasticsearch-operator-uninstall:
 	olm_deploy/scripts/operator-uninstall.sh
-
+gen-dockerfiles:
+	./hack/generate-dockerfile-from-midstream > Dockerfile && \
```
Comment (ewolinetz) [inline Makefile:236]: this seems like its missing the `Dockerfile.in` parameter

Comment (jcantrill) [reply Makefile:236]: It assumes defaults, which can be overriden

File: Makefile [hunk 5]
```diff
@@ -85,12 +85,8 @@ lint-dockerfile:
 	@hack/lint-dockerfile
 .PHONY: lint-dockerfile
 
-.INTERMEDIATE: Dockerfile.dev
-Dockerfile.dev: Dockerfile Dockerfile.centos.patch
-	patch -o Dockerfile.dev Dockerfile Dockerfile.centos.patch
-
 image: .output/image
-.output/image: Dockerfile.dev $(GO_FILES) $(MANIFEST_FILES) $(OTHER_FILES)
+.output/image: $(GO_FILES) $(MANIFEST_FILES) $(OTHER_FILES)
```
Comment (ewolinetz) [inline Makefile:89]: can you add the dockerfile generation as a requirement for this?

Comment (jcantrill) [reply Makefile]: updated

**Review by ewolinetz — APPROVED** [review]

**Pull request merged**